interpreter/cling/lib/Interpreter/BackendPasses.cpp

```diff
@@ -382,56 +353,44 @@ void BackendPasses::CreatePasses(llvm::Module& M, int OptLevel)
-
-  PMBuilder.DisableUnrollLoops = !m_CGOpts.UnrollLoops;
-  PMBuilder.MergeFunctions = m_CGOpts.MergeFunctions;
-
-  PMBuilder.LibraryInfo = new TargetLibraryInfoImpl(m_TM.getTargetTriple());
+  PassBuilder PB(&m_TM);
 
   // At O0 and O1 we only run the always inliner which is more efficient. At
   // higher optimization levels we run the normal inliner.
-  // See also call to `CGOpts.setInlining()` in CIFactory!
-  if (PMBuilder.OptLevel <= 1) {
-    bool InsertLifetimeIntrinsics = PMBuilder.OptLevel != 0;
-    PMBuilder.Inliner = createAlwaysInlinerLegacyPass(InsertLifetimeIntrinsics);
+  if (OptLevel <= 1) {
+    m_MPM.addPass(AlwaysInlinerPass());
   } else {
-    PMBuilder.Inliner = createFunctionInliningPass(OptLevel,
-                                                   PMBuilder.SizeLevel,
-            (!m_CGOpts.SampleProfileFile.empty() && m_CGOpts.PrepareForThinLTO));
-  }
+    // Otherwise, use the default pass pipeline. We also have to map our
+    // optimization levels into one of the distinct levels used to configure
+    // the pipeline.
+    OptimizationLevel Level = mapToLevel(m_CGOpts);
 
-  // Set up the per-module pass manager.
-  m_MPM[OptLevel].reset(new legacy::PassManager());
+    m_MPM = PB.buildPerModuleDefaultPipeline(Level);
+  }
 
-  m_MPM[OptLevel]->add(new KeepLocalGVPass());
-  m_MPM[OptLevel]->add(new PreventLocalOptPass());
-  m_MPM[OptLevel]->add(new WeakTypeinfoVTablePass());
-  m_MPM[OptLevel]->add(new ReuseExistingWeakSymbols(m_JIT));
+  m_MPM.addPass(KeepLocalGVPass());
+  m_MPM.addPass(PreventLocalOptPass());
+  m_MPM.addPass(WeakTypeinfoVTablePass());
+  m_MPM.addPass(ReuseExistingWeakSymbols(m_JIT));
 
   // The function __cuda_module_ctor and __cuda_module_dtor will just generated,
   // if a CUDA fatbinary file exist. Without file path there is no need for the
   // function pass.
   if(!m_CGOpts.CudaGpuBinaryFileName.empty())
-    m_MPM[OptLevel]->add(new UniqueCUDAStructorName());
-  m_MPM[OptLevel]->add(createTargetTransformInfoWrapperPass(
-                                                   m_TM.getTargetIRAnalysis()));
+    m_MPM.addPass(UniqueCUDAStructorName());
+
+  // Register all the basic analyses with the managers.
+  PB.registerModuleAnalyses(MAM);
+  PB.registerCGSCCAnalyses(CGAM);
+  PB.registerFunctionAnalyses(FAM);
+  PB.registerLoopAnalyses(LAM);
+  PB.crossRegisterProxies(LAM, FAM, CGAM, MAM);
 
   //if (!CGOpts.RewriteMapFiles.empty())
   //  addSymbolRewriterPass(CGOpts, m_MPM);
 
-  PMBuilder.populateModulePassManager(*m_MPM[OptLevel]);
-
-  m_FPM[OptLevel].reset(new legacy::FunctionPassManager(&M));
-  m_FPM[OptLevel]->add(createTargetTransformInfoWrapperPass(
-                                                   m_TM.getTargetIRAnalysis()));
   if (m_CGOpts.VerifyModule)
-      m_FPM[OptLevel]->add(createVerifierPass());
-  PMBuilder.populateFunctionPassManager(*m_FPM[OptLevel]);
+    m_MPM.addPass(VerifierPass());
```

Comment:
The verifier pass should be earlier, at the very least before the default passes